### PR TITLE
[Wallet] Add support for -walletnotify to execute custom shell tasks for incom…

### LINKED ISSUE #3016: Implement `walletnotify` configuration parameter
To match Bitcoin API I would like to see this feature implemented

### PR DESCRIPTION
**Pull request by sondreb**
Closes #3016

- Add a shell helper to execute shell commands.
- Add a new option to configure the wallet notify.

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.Wallet.Tests/WalletManagerTest.cs [hunk 1]
```diff
@@ -33,6 +33,92 @@ public WalletManagerTest(WalletFixture walletFixture)
             this.walletFixture = walletFixture;
         }
 
+        /// <summary>
+        /// Creates a wallet with valid trx and verifies that the ShellHelper is called for the found UTXO's.
+        /// </summary>
+		[Fact]
+		public void VerifyExecutionOfWalletNotify()
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet.Tests/WalletManagerTest.cs]: @sondreb it looks like all the code changes were submitted with extra indentation?

**Comment by fassadlr** [comment]
Fix conflicts please :)

File: src/Stratis.Bitcoin/Utilities/ShellHelper.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+                Arguments = args,
+                RedirectStandardOutput = wait,
+                RedirectStandardError = wait,
+                UseShellExecute = false,
+                CreateNoWindow = true,
+            };
+
+            if (!wait)
+            {
+                Process.Start(startInfo);
```
Comment (fassadlr) [inline src/Stratis.Bitcoin/Utilities/ShellHelper.cs]: Arent we probably exposing ourselves here to a security issues? We can then just execute any code here?

**Comment by sondreb** [comment]
The code is not based on the Bitcoin implementation, other than in the nature of how it works. Here are the interesting parts from Bitcoin source that perhaps the PR should be updated with:

Here is a repo search in the bitcoin source on "walletnotify": https://github.com/bitcoin/bitcoin/search?utf8=%E2%9C%93&q=walletnotify&type=

Argument comment could maybe be copied?

`gArgs.AddArg("-walletnotify=<cmd>", "Execute command when a wallet transaction changes (%s in cmd is replaced by TxID)", false, OptionsCategory::WALLET);`

wallet.ccp:

```
    // notify an external script when a wallet transaction comes in or is updated
    std::string strCmd = gArgs.GetArg("-walletnotify", "");

    if (!strCmd.empty())
    {
        boost::replace_all(strCmd, "%s", wtxIn.GetHash().GetHex());
        std::thread t(runCommand, strCmd);
        t.detach(); // thread runs free
    }
```

So the execution of the thread is simpler in C++, it does basically the same by not waiting for the new thread to complete.

Could consider removing the additional support for waiting for exit in the ShellHelper, to not have code that is not actually used yet.

**Comment by dangershony** [comment]
@sondreb any chance this can be a standalone feature that depend on wallet?

**Comment by mikedennis** [comment]
Added details from bitcoin implementation:
```
void runCommand(const std::string& strCommand)
{
    if (strCommand.empty()) return;
#ifndef WIN32
    int nErr = ::system(strCommand.c_str());
#else
    int nErr = ::_wsystem(std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>,wchar_t>().from_bytes(strCommand).c_str());
#endif
    if (nErr)
        LogPrintf("runCommand error: system(%s) returned %d\n", strCommand, nErr);
}
```

**Comment by sondreb** [comment]
> @sondreb any chance this can be a standalone feature that depend on wallet?

I will try to refactor it and see, and thanks to @mikedennis for the Bitcoin explain, I'll make sure it is even more similar in an update.

**Comment by sondreb** [comment]
I'm refactoring this PR to utilize the new EventHub implementation, but I have something that must be clarified:

- Should this feature be "standalone" and loaded in the daemon using .UseWalletNotify()
- Should WalletFeature depends on the WalletNotifyFeature? If so, then the "FoundTransaction" event must be stored globally, not in the Wallet feature.

I'm leaning towards making the WalletNotifyFeature dependent on Wallet, there is no other dependency between WalletNotifyFeature and WalletFeature than the event class at the moment.

**Comment by fassadlr** [comment]
@sondreb @dangershony What is the status of this PR? Can this be closed or is still in active development?

**Comment by dangershony** [comment]
I think you can close its already part of a repo with collections of features
https://github.com/block-core/blockcore-features/tree/master/Blockcore.Features.WalletNotify

**Comment by fassadlr** [comment]
Thank @dangershony !

**Pull request closed without merging**